### ISSUE TITLE
Querier chunk: projected schema interning/caching

### ISSUE THREAD

**Issue by crepererum** [influxdata/influxdb_iox#5202]
While we cache and re-use table-level schemas (shared via an `Arc`), we currently produce a new `Schema` object for every querier chunk due to projection:

https://github.com/influxdata/influxdb_iox/blob/614ca5ca96f0df7fea90b611122dbad0525d056e/querier/src/chunk/mod.rs#L488-L492

In prod we have some schemas w/ >600 columns. Every columns requires at least 3 strings:

1. The column name
2. The per-column metadata key for the IOx data type
3. The per-column metadata value for the IOx data type

We also have one table w/ over 30k chunks (probably due to compaction issue, but let's not play whataboutism now). That adds up to multiple GBs of chunk-level metadata:

![Screenshot from 2022-07-25 13-23-33](https://user-images.githubusercontent.com/1529400/180810134-4f85edda-f937-4a2c-a0e2-a6760266ad47.png)

I propose to add a new `ProjectedSchema` cache for that:

```rust
Cache <
    K = (TableId, Vec<ColumnName>),
    V = Arc<Schema>,  // projected schema
    GetExtra = Arc<Schema>,  // unprojected schema
    PeekExtra = (),  // no peek
>
```

The cache loader is basically `Arc::new(schema.select_by_names(...))` (see code linked above).

**Comment by alamb** [comment]
"whataboutism" --love it

**Comment by alamb** [comment]
For what it is worth, I suggest adding some sort of self preservation check that doesn't even try to plan a query with 30k chunks -- it isn't likely going to work at all, even if you fix this particular problem.

Not that this ticket's suggestion isn't valuable (it is!) I am just reiterating out fixing this will likely still not allow the query with 30k files to actually complete

**Comment by crepererum** [comment]
> For what it is worth, I suggest adding some sort of self preservation check that doesn't even try to plan a query with 30k chunks -- it isn't likely going to work at all, even if you fix this particular problem.

I thought about this again and still agree with the self-preservation, but the check is hard to implement without the change proposed in this ticket because we probably wanna reject the query if there are too many chunks after the query predicates are applied (= filtering via chunk stats), not if there are too many chunks in the table in general. Otherwise the prod querier would currently reject about every query.

Filtering (for the sake of simplicity and effort) is implemented using the same chunk type that the query engine uses, so we must at least be able to create them (even when we filter them out before actually using their payload within the query). But we're probably not even able to do that under all circumstances due to the issue described in this ticket.

**Comment by crepererum** [comment]
Closed (by #5215).
